packages/plugin-drizzle/src/storage.ts

```diff
@@ -0,0 +1,259 @@
+});
+
+export type ReorgRollbackRow = typeof reorgRollbackTable.$inferSelect;
+
+export async function initializeReorgRollbackTable<
+  TQueryResult extends PgQueryResultHKT,
+  TFullSchema extends Record<string, unknown> = Record<string, never>,
+  TSchema extends
+    TablesRelationalConfig = ExtractTablesWithRelations<TFullSchema>,
+>(tx: PgTransaction<TQueryResult, TFullSchema, TSchema>) {
+  try {
+    // Create the audit log table
+    await tx.execute(
+      sql.raw(`
+        CREATE TABLE IF NOT EXISTS __reorg_rollback(
+          n SERIAL PRIMARY KEY,
+          op CHAR(1) NOT NULL,
+          table_name TEXT NOT NULL,
+          cursor INTEGER NOT NULL,
+          row_id TEXT,
+          row_value JSONB
+        );
+      `),
+    );
+
+    // Create the trigger function
+    await tx.execute(
+      sql.raw(`
+        CREATE OR REPLACE FUNCTION reorg_checkpoint()
+        RETURNS TRIGGER AS $$
+        DECLARE
+          id_col TEXT := TG_ARGV[0]::TEXT;
+          order_key INTEGER := TG_ARGV[1]::INTEGER;
+          new_id_value TEXT := row_to_json(NEW.*)->>id_col;
+          old_id_value TEXT := row_to_json(OLD.*)->>id_col;
+        BEGIN
+          IF (TG_OP = 'DELETE') THEN
+            INSERT INTO __reorg_rollback(op, table_name, cursor, row_id, row_value)
+              SELECT 'D', TG_TABLE_NAME, order_key, old_id_value, row_to_json(OLD.*);
+          ELSIF (TG_OP = 'UPDATE') THEN
+            INSERT INTO __reorg_rollback(op, table_name, cursor, row_id, row_value)
+              SELECT 'U', TG_TABLE_NAME, order_key, new_id_value, row_to_json(OLD.*);
+          ELSIF (TG_OP = 'INSERT') THEN
+            INSERT INTO __reorg_rollback(op, table_name, cursor, row_id, row_value)
+              SELECT 'I', TG_TABLE_NAME, order_key, new_id_value, null;
+          END IF;
+          RETURN NULL;
+        END;
+        $$ LANGUAGE plpgsql;
+      `),
+    );
+  } catch (error) {
+    console.error(error);
+    throw new DrizzleStorageError(
+      `Failed to initialize reorg rollback table: ${error}`,
+    );
+  }
+}
+
+export async function registerTriggers<
+  TQueryResult extends PgQueryResultHKT,
+  TFullSchema extends Record<string, unknown> = Record<string, never>,
+  TSchema extends
+    TablesRelationalConfig = ExtractTablesWithRelations<TFullSchema>,
+>(
+  tx: PgTransaction<TQueryResult, TFullSchema, TSchema>,
+  tables: string[],
+  endCursor: Cursor,
+  idColumn: string,
+) {
+  try {
+    for (const table of tables) {
+      await tx.execute(
+        sql.raw(`DROP TRIGGER IF EXISTS ${table}_reorg ON ${table};`),
+      );
+      await tx.execute(
+        sql.raw(`
+          CREATE CONSTRAINT TRIGGER ${table}_reorg
+          AFTER INSERT OR UPDATE OR DELETE ON ${table}
+          DEFERRABLE INITIALLY DEFERRED
+          FOR EACH ROW EXECUTE FUNCTION reorg_checkpoint('${idColumn}', ${`${Number(endCursor.orderKey)}`});
+        `),
+      );
+    }
+  } catch (error) {
+    console.error(error);
```

Comment:
We should not log the error like this. We can add a `drizzle: Error` property to `DrizzleStorageError` and change the constructor to be like `new DrizzleStorageError("mesage...", { error })`